patches/api/0446-Add-FluidState-API.patch

```diff
@@ -0,0 +1,171 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: vicisacat <victor.branchu@gmail.com>
+Date: Fri, 17 Nov 2023 20:21:47 +0100
+Subject: [PATCH] Add FluidState API
+
+
+diff --git a/src/main/java/io/papermc/paper/block/fluid/FluidData.java b/src/main/java/io/papermc/paper/block/fluid/FluidData.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..ece4b71a5882d43f8149e559c746f4dba56648e1
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/block/fluid/FluidData.java
+@@ -0,0 +1,75 @@
++package io.papermc.paper.block.fluid;
++
++import io.papermc.paper.math.Position;
++import org.bukkit.Fluid;
++import org.bukkit.Location;
++import org.bukkit.World;
++import org.bukkit.util.Vector;
++import org.jetbrains.annotations.NotNull;
++
++public interface FluidData extends Cloneable {
++
++    /**
++     * Gets the fluid type of this fluid data.
++     *
++     * @return the fluid type
++     */
++    @NotNull Fluid getFluidType();
++
++    /**
++     * Returns a copy of this FluidData.
++     *
++     * @return a copy of the fluid data
++     */
++    @NotNull FluidData clone();
++
++    /**
++     * Returns the flowing direction the liquid as a vector
++     *
++     * @return the flow direction vector
++     */
++    Vector getFlowDirection(Location location);
++
++    /**
++     * Returns the amount of liquid is in the fluid, as in the number of layers.
++     *
++     * @return the amount as an integer, between 0 and 8
++     */
++    int getAmount();
++
++    /**
++     * Returns the height of the fluid in blocks
++     *
++     * @return the height as a float value, representing the height in blocks
++     */
++    float getHeight();
++
++    /**
++     * Returns whether this fluid is the same fluid as the one passed in the parameter
++     *
++     * @param fluid a fluid
++     * @return true if they are the same, false otherwise
++     */
++    boolean is(Fluid fluid);
++
++    /**
++     * Returns whether this fluid is a source block
++     *
++     * @return true if the fluid is a source block, false otherwise
++     */
++    boolean isSource();
++
++    /**
++     * Returns whether this fluid is water.
++     *
++     * @return true if the fluid is water, false otherwise
++     */
++    boolean isWater();
++
++    /**
++     * Returns whether this fluid is lava.
++     *
++     * @return true if the fluid is lava, false otherwise
++     */
++    boolean isLava();
```

Comment:
something along the lines of `Fluid.isWater(Fluid fluid)`?

Reply:
Yes, (assuming not static here)

Reply:
Do note here that this may be better to be covered FluidTag in future, but this would be future stuff, as the tag API won't work nicely with fluids since it's an enum currently.

Reply:
Is this good or needlessly excessive ?
<details>
  <summary>Code</summary>
  
```java
public enum Fluid implements Keyed {
   // ...
    WATER{
        @Override
        public boolean isWater() {
            return true;
        }
    },
    /**
     * Flowing water.
     */
    FLOWING_WATER{
        @Override
        public boolean isWater() {
            return true;
        }
    },
    /**
     * Stationary lava.
     */
    LAVA{
        @Override
        public boolean isLava() {
            return true;
        }
    },
    // ...
}
// ...
public boolean isWater() {
        return false;
    }
    
    public boolean isLava() {
        return false;
    }
```
</details>